### ISSUE TITLE
Relative state depth issue

### ISSUE THREAD

**Issue by calekennedy** [reactjs/react-redux#411]
I am having a little bit of trouble generalizing the relative depth of my Containers within the redux store.

For instance, lets say I have a Container that uses the function 
```
import _ from 'lodash';

const mapStateToProps = (state) => {
  const country = _.get(state, 'Address.Country', {});
  const { value } = country;
  return { value };
};
```

This is fine, for now, because I'm calling 
```
combineReducers({
   /*...*/, 
  Country: Country.Reducer,
   /*...*/
});
```
inside of my Address reducer, and in my Root reducer, I am doing the same with Address.

Now, let's imagine that I create a deeper nesting. Lets say I have a store that looks more like 
```
{ User: { Address: { Country: { value:  'USA' }  }  }  }
``` 
Do I need to inject the new relative path into the Country container? 
```
_.get(state, 'Address.Country', {}); // undefined
_.get(state, 'User.Address.Country', {}); // new relative path
```

Or am I missing something?

Basically, I'd like the containers to be depth-agnostic. My reducers already operate this way, I don't understand why containers are necessarily different. 

Perhaps there could be a way to receive both relative state and global state within a mapStateToProps call?

**Comment by markerikson** [comment]
This is a usage question, and should really be asked over on Stack Overflow.

That said: reducers are frequently structured so that they only know about the scope of the data they need to update, using the `combineReducers` utility.  However, any given component _could_ be interested in any portion of the entire state tree.  Also, Redux only stores a reference to the state tree as a whole anyway, so that's what's returned by `getState()`.

This is where "selector" functions come in.  Selectors generally receive the entire state as an argument, and extract some small piece out of the whole.  That way a component doesn't need to know what the entire state structure looks like - the selector is abstracting it.  Think of selectors as an "SQL-like query" for a state tree.

The Redux docs talk about selectors in http://redux.js.org/docs/recipes/ComputingDerivedData.html, and Dan discusses them in his recent "Idiomatic Redux" video series: http://redux.js.org/docs/recipes/ComputingDerivedData.html.
